pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+                    'Ack request failed, better luck next batch', exc_info=e)
+                continue
+
+            if metrics_client:
+                metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _fire_callback(semaphore: asyncio.Semaphore,
+                             message: SubscriberMessage,
+                             callback: ApplicationHandler,
+                             ack_queue: 'asyncio.Queue[str]',
+                             metrics_client: Optional[MetricsAgent]
+                             ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.succeeded')
+                metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                         time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            if metrics_client:
+                metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: Optional[MetricsAgent]) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                if metrics_client:
+                    metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            if metrics_client:
+                metrics_client.histogram(
+                    'pubsub.consumer.latency.receive',
+                    time.time() - message.publish_time.timestamp())
+
+            asyncio.ensure_future(_fire_callback(
+                semaphore,
+                message,
+                callback,
+                ack_queue,
+                metrics_client,
+            ))
+            message_queue.task_done()
+
+    async def producer(
+            subscription: str,
+            message_queue: MessageQueue,
+            subscriber_client: 'SubscriberClient',
+            max_messages: int,
+            metrics_client: Optional[MetricsAgent]
+    ) -> None:
+        while True:
+            try:
+                new_messages = await subscriber_client.pull(
+                    subscription=subscription, max_messages=max_messages)
+            except asyncio.TimeoutError:
```

Comment:
Hmm, but what is it for? `TimeoutError` is thrown when there was no messages to pull for ~10 secs. In a worst case scenario when we have no message flow we'll be 100% ready to immediately consume messages for a price of 6 RPM * number of workers load. And we probably don't care about stressing google's api anyway?

Reply:
I wasn't quite thinking of the 10s thing -- I wonder if we should explicitly set the `timeout=10` arg to `pull()` here to make sure it's adjacent to the `except TimeoutError`... that might make it more clear when this case would occur.

I guess the case I was thinking about was mostly the permafailure thing: though your points are correct, if Google's APIs are down we might want to have a louder failure than everything running smoothly here and just-so-happening to be looping early without successfully grabbing messages. Users can solve this with alerts one their processing volume, but... well, if your subscriber is repeatedly failing to grab messages from the subscription, that feels like it should be a really loud error to me. If this happens a large amount of times in a row, we're basically in a "literally everything is non-functional" state.

Reply:
Hmm, let me clarify this a bit. `TimeoutError` here is a part of _normal_ flow. That's just how pull works.
Apart from that we don't catch any exception. That means that for any actual error (say, 500) the worker will throw -> the exception will surface -> app will crash and hopefully be restarted by kubernetes.

So if anything I'd say we need backoff here to not fail too loud immediately.

There's also the case when google is so down we're getting `TimeoutErrors` from it. This way we can't distinguish failures from no messages. Not sure how to go about this 🤔

Reply:
Ah, yeah, distinguishing the normal case from failure modes is where I got hung up. If Google returns Timeout errors when there are no message to-be-processed, your code looks 100% correct. If there's a chance of those Timeouts being due to API issues / networking problems / whatever, that's where I was thinking we needed more explicit error handling.

I'm not actually sure how to differentiate or react differently... are the payloads any different? Any thoughts from other @talkiq/engineering   folks?

Reply:
Is there seriously no way to distinguish between a network timeout and no messages in subscription -____-

if this is an concern would it be too much overhead to ping the server when a timeout happens to see if its a network timeout? or does pubsub support some sort of `echo` or `healthcheck` type thing

Reply:
I wonder if this is a bad thing at all. Imagine we _could_ tell that google is down? If we crash the app it will only increase our time to recovery, since google might be back any second but redeploy cycle takes minutes. Thoughts?

Reply:
I mean if there are no messages to pull the response should be an empty list of messages and not a `TimeoutError` right?

https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull#response-body

> receivedMessages[] -> Received Pub/Sub messages. The list will be empty if there are no more messages available in the backlog. For JSON, the response can be entirely empty. The Pub/Sub system may return fewer than the maxMessages requested even if there are more messages available in the backlog.

Reply:
So in the case where there are no messages we go through `for m in messages` and do nothing, then `await queue.join()` and cycle right back to a new request over and over right? I'm sure that in that scenario the biggest thing we have to worry about is getting rate limited.

Reply:
@jonathan-johnston my understanding is that returning zero messages from that endpoint might be depreacated -- see `returnImmediately` on that page, which used to be the way to say "return immediately if there are less than `maxMessages` items available instead of waiting for some arbitrary amount of time.

...that said, I realize that value is deprecated, but it sure sounds like it might be what we want. "Otherwise, the system may wait (for a bounded amount of time) until at least one message is available, rather than returning no messages." scares me -- this line might have arbitrary latency values which we can't control? Is there any way to set that timeout value? This might be the reason why we're needing multiple producers right now, since each of them is arbitrarily waiting sometimes...

Reply:
@TheKevJames why this scares you tho? I mean if we trust google that request will only wait until new messages are available it should be ok. If we don't trust google... we'll retry in `session timeout` seconds anyway?
So how it works right now -> we send `pull` request and, if the queue is empty, just hang there for 10 seconds (default session timeout in `SubscriberClient`. If new messages available - cool. If not - let's try again.

Reply:
It's mostly the "trust Google" part that worries me haha. We're relying here on their idea of "bounded amount of time" being reasonable and matching our assumptions. See for example how aiohttp's default idea of a "reasonable default timeout" was something like ten minutes, which makes absolutely no sense in a lot of the places we use it.

What happens if Google decides "reasonable" is "24 hours"? What if the http connection breaks in a way what our client can't detect, so we sit waiting for them to timeout the connection but never get that notification? Could this lead to zombie threads sitting around doing nothing?

Reply:
Our default timeout is `10s` and totally overridable by user. We never hang on a line for more than 10 seconds by default.